### PR TITLE
Turn fold/unfold statements into assert statements.

### LINKED ISSUE #18: Investigate partial predicate accesses in `fold`/`unfold`/`unfolding`
When #10 is done, all `fold` and `unfold` statements will be eliminated, and `unfolding` statements will be replaced by their bodies. For instance,

```js
fold acc(tuple(this), 1/2)
unfolding acc(tuple(this), 1/2) in ...
unfold acc(tuple(this), 1/2)
```

will all disappear from method bodies. However, we might not want to do this depending on the access value (`1/2` above). Preliminary work would include:
1. Trying to come up with counterexamples where this behaviour causes previously-verifying code to no longer verify.
2. If there are counterexamples, coming up with how these statements should be handled to avoid them.

### PR DESCRIPTION
**Pull request by ionathanch**
Haven't tested this thoroughly yet, and this is still just a conjecture, but here's a PR anyway (will add tests tmrw).

In the tutorial, it states:

> An unfold operation exchanges a predicate instance for its body; roughly speaking, the predicate instance is exhaled, and its body inhaled. Such an operation causes a verification failure if the predicate instance is not held. A fold operation exchanges a predicate body for a predicate instance; roughly speaking, the body is exhaled, and the instance inhaled.

Since inlining replaces predicate permission obtainment by permission obtainment of its body, and similarly for their release, it's essentially as if the body has been inhaled (similarly, exhaled). By asserting the bodies where folds/unfolds occur, we say no more than that we have the permissions that we _already should_, along with whatever other assertions that live in the predicate body. That should solve the errors in [this comment](https://github.com/jyoo980/silver/issues/34#issuecomment-748754970).

As a bonus, we no longer simply discard any partial predicate access values (which was suspicious to me in the first place), since they are instead asserted. This maybe could resolve #18.

### CODE REVIEW

File: src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala [hunk 1]
```diff
@@ -92,20 +92,24 @@ trait InlineRewrite extends PredicateExpansion with InlineErrorChecker {
...
+    ViperStrategy.CustomContext[Set[String]]({
+      case (fold@Fold(PredicateAccessPredicate(pred, perm)), ctxt) =>
+        if (cond(pred.predicateName)) {
+          val optPredbody = propagatePermission(pred.predicateBody(program, ctxt), perm)
+          (Assert(optPredbody.get)(fold.pos, fold.info, fold.errT), ctxt)
+        } else (fold, ctxt)
+      case (unfold@Unfold(PredicateAccessPredicate(pred, perm)), ctxt) =>
+        if (cond(pred.predicateName)) {
+          val optPredbody = propagatePermission(pred.predicateBody(program, ctxt), perm)
+          (Assert(optPredbody.get)(unfold.pos, unfold.info, unfold.errT), ctxt)
```
Comment (jyoo980) [inline src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala:111]: Should be safe to call `.get` here right

**Review by jyoo980 — APPROVED** [review]

**Comment by braxtonhall** [comment]
I've tested this on a few small examples and it seems to be working. Gonna try the big test suite now

File: src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala (hunk 1 above)
Comment (ionathanch) [reply src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala:111]: Yeah, since `cond` now checks that predicates are defined first

**Comment by ionathanch** [comment]
This table http://viper.ethz.ch/tutorial/?page=1&section=#assertion-checking seems to suggest that unfolds (analogous to inhales) should be replaced by assumes, and folds (analogous to exhales) by asserts, but `assume` can't have any access predicates in it, and I think my original reasoning is still sound

File: src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala (hunk 1 above)
Comment (ionathanch) [reply src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala:111]: Oh that wasn't a question
I do `.get` on line 111

Comment (jyoo980) [reply src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala:111]: Oh It was a question, I'm just shitty at asking them online ty

**Comment by ionathanch** [comment]
Tested on this example:
```javascript
field a: Int
field b: Int
field c: Int

predicate A(this: Ref) {
  acc(this.a) && this.a > 0
}

predicate B(this: Ref) {
  acc(this.b) && A(this) && this.b > 0
}

predicate C(this: Ref) {
  acc(this.c) && B(this) && this.c > 0
}

method example(this: Ref)
  requires C(this)
  ensures  C(this)
{
  unfold C(this)
  unfold B(this)
  unfold A(this)
  this.a := this.b + this.c
  this.b := this.c
  fold A(this)
  fold B(this)
  fold C(this)
}
```
Yields this method:
```javascript
method example(this: Ref)
  requires acc(this.c, write * write) && (acc(this.b, write * (write * write)) && (acc(this.a, write * (write * (write * write))) && this.a > 0 && this.b > 0) && this.c > 0)
  ensures acc(this.c, write * write) && (acc(this.b, write * (write * write)) && (acc(this.a, write * (write * (write * write))) && this.a > 0 && this.b > 0) && this.c > 0)
{
  assert acc(this.c, write * write) && (acc(this.b, write * (write * write)) && (acc(this.a, write * (write * (write * write))) && this.a > 0 && this.b > 0) && this.c > 0)
  assert acc(this.b, write * write) && (acc(this.a, write * (write * write)) && this.a > 0 && this.b > 0)
  assert acc(this.a, write * write) && this.a > 0
  this.a := this.b + this.c
  this.b := this.c
  assert acc(this.a, write * write) && this.a > 0
  assert acc(this.b, write * write) && (acc(this.a, write * (write * write)) && this.a > 0 && this.b > 0)
  assert acc(this.c, write * write) && (acc(this.b, write * (write * write)) && (acc(this.a, write * (write * (write * write))) && this.a > 0 && this.b > 0) && this.c > 0)
}
```

**Review by braxtonhall — APPROVED** [review]
Ran the test suite (up until it hangs lol)
Across the 44 failing tests none of them are _new_.
Some of the tests fail for a different reason but are closer to the test suite's definition of "correct"

**Comment by ionathanch** [comment]
One more thing: `unfolding` isn't being handled here. I don't think there's any "assertion expressions", only assertion statements, so the same thing can't really be done. If there's nothing obvious that can be done I'll just mention it as future work

**Comment by ionathanch** [comment]
Quick demonstration of what can go wrong with `unfolding`:
```javascript
predicate P(this: Ref) { true }

method test(this: Ref)
{
  assert unfolding P(this) in true
}
```
This does not pass verification, since `test` doesn't have permission to unfold `P(this)`. However, this gets "expanded" into:
```javascript
predicate P(this: Ref) { true }

method test(this: Ref)
{
  assert true
}
```
which _does_ pass verification, which is wrong.

EDIT: Actually that happens for unfold/fold too. `unfold P(this)` in the body (failing) would turn into `assert true` (succeeding). I guess any failures relying on the presence of an unjustified fold/unfold/unfolding ceases to fail after our transformation. Not related to this PR (oops), but it would be an interesting Future Work to discuss. It seems to me more like a shortcoming of the design of predicates in Viper than anything lol

**Review by alison-li — APPROVED** [review]

**Comment by ionathanch** [comment]
After looking at some of the failing tests that have to do with the disappearance of `fold`/`unfold` and even `unfolding`, I'm not so sure this is 100% the correct solution, but given that a bunch of tests _are_ passing and my original reasoning, I'm going to merge it in anyway, and I'll talk about the downsides in evaluation or future work or smth.

**Pull request merged**